### ISSUE TITLE
Slow shutdown due to plugin termination problems?

### ISSUE THREAD

**Issue by towolf** [mattermost/mattermost-server#10060]
#### Summary

Shutdown of Mattermost 5.6.2 is slow. Takes 90 seconds to terminate.

#### Steps to reproduce

Enable some plugins, then run `systemctl restart mattermost`

Observe logs:
```
2019-01-03T13:42:26.574+0100	info	jobs/schedulers.go:140	Stopping schedulers.
2019-01-03T13:42:26.574+0100	info	jobs/schedulers.go:75	Schedulers stopped.
2019-01-03T13:42:26.574+0100	info	jobs/workers.go:176	Stopped workers
2019-01-03T13:42:26.574+0100	info	app/app.go:216	Stopping Server...
2019-01-03T13:42:26.574+0100	info	app/web_hub.go:120	stopping websocket hub connections
<wait here>
2019-01-03T13:43:57.052+0100	info	utils/i18n.go:78	Loaded system translations for 'en' from '/opt/mattermost/i18n/en.json'
2019-01-03T13:43:57.053+0100	info	app/app.go:143	Server is initializing...
2019-01-03T13:43:57.057+0100	info	sqlstore/supplier.go:207	Pinging SQL master database
2019-01-03T13:43:57.160+0100	info	commands/server.go:82	Current version is 5.6.0 (5.6.2/Sat Dec 22 14:58:39 UTC 2018/3aed6efc6bb5a67fdb8237aac38d46fab3266e8d/2c2d33b22b4c2938938eade4b9914748c1be9d3b)
```
Next time there are weird messages related to terminating plugins in the logs
```
2019-01-03T14:04:01.534+0100	info	jobs/schedulers.go:140	Stopping schedulers.
2019-01-03T14:04:01.534+0100	info	jobs/schedulers.go:75	Schedulers stopped.
2019-01-03T14:04:05.969+0100	info	jobs/workers.go:176	Stopped workers
2019-01-03T14:04:05.970+0100	info	app/app.go:216	Stopping Server...
2019-01-03T14:04:05.970+0100	info	app/web_hub.go:120	stopping websocket hub connections
2019-01-03T14:04:05.977+0100	info	app/plugin.go:193	Shutting down plugins
2019-01-03T14:04:05.978+0100	warn	go-plugin/client.go:384	error closing client during Kill	{"plugin_id": "jira", "wrapped_extras": "errconnection is shut down"}
2019-01-03T14:04:05.978+0100	warn	go-plugin/client.go:384	error closing client during Kill	{"plugin_id": "mattermost-autolink", "wrapped_extras": "errconnection is shut down"}
2019-01-03T13:59:00.772+0100	info	sqlstore/supplier.go:850	Closing SqlStore
2019-01-03T13:59:00.774+0100	info	app/app.go:237	Server stopped
2019-01-03T13:59:01.287+0100	info	utils/i18n.go:78	Loaded system translations for 'en' from '/opt/mattermost/i18n/en.json'
2019-01-03T13:59:01.288+0100	info	app/app.go:143	Server is initializing...
2019-01-03T13:59:01.292+0100	info	sqlstore/supplier.go:207	Pinging SQL master database
2019-01-03T13:59:01.401+0100	info	commands/server.go:82	Current version is 5.6.0 (5.6.2/Sat Dec 22 14:58:39 UTC 2018/3aed6efc6bb5a67fdb8237aac38d46fab3266e8d/2c2d33b22b4c2938938eade4b9914748c1be9d3b)

```

#### Expected behavior
Quick shutdown and restart

#### Observed behavior (that appears unintentional) 
See above

#### Possible fixes
Just a hunch:

Maybe `mattermost.service` should contain [KillMode=process](https://www.freedesktop.org/software/systemd/man/systemd.kill.html#KillMode=)?
 
If Mattermost wants to kill the plugins itself, then maybe systemd should keep its hand off?

**Comment by amyblais** [comment]
Hi @towolf, letting you know that I've asked our team about this and I'll let you know once I hear their feedback.

**Comment by lieut-data** [comment]
Given the delay occurs just after the message `stopping websocket hub connections`, it's possible this is an artifact of my changes in https://github.com/mattermost/mattermost-server/pull/9863, though nothing jumps out as a cause yet.

@towolf, can you verify the timestamps in your second log entry? They seem curiously out of order:
```
2019-01-03T14:04:01.534+0100
...
2019-01-03T14:04:05.969+0100
...
2019-01-03T13:59:00.772+0100
```

Does this problem reproduce when plugins are disabled? If not, can you share what plugins you have enabled and and relevant configuration to help us reproduce?

**Comment by towolf** [comment]
> They seem curiously out of order

Right, not sure what I did there. I think I puzzled these together somehow.

The thing is, that every other time, the logs would not print before `mattermost` process was getting KILL (90s timeout).

And then on the next time it printed something about terminating plugins. 

I added the mentioned systemmd config `KillMode=process` now, and shutdown is always immediate.

> If not, can you share what plugins you have enabled and and relevant configuration to help us reproduce?

Don't want to disable the plugins now, would interfere with live operations ...

Plugin config:
```json
    "PluginSettings": {
        "Enable": true,
        "EnableUploads": false,
        "Directory": "/var/lib/mattermost/plugins",
        "ClientDirectory": "/var/lib/mattermost/client/plugins",
        "Plugins": {
            "jira": {
                "enabled": true,
                "secret": "eaX... ... ...qZoo",
                "username": "system-bot"
            },
            "mattermost-autolink": {
                "links": [
                    {
                        "Pattern": "(?P<jira_id>[A-Z]{2,16}-\\d{2,6})",
                        "Template": "[${jira_id}](https://jira.example.com/browse/${jira_id})"
                    },
                    {
                        "Pattern": "k8s-prod-logs:(?P<search_term>.*)",
                        "Template": "[k8s-prod-logs:${search_term}](https://logs.example.com/search?relative=86400&q=${search_term})"
                    },
                    {
                        "Pattern": "k8s-stage-logs:(?P<search_term>.*)",
                        "Template": "[k8s-stage-logs:${search_term}](https://logs.example.com/search?relative=86400&q=${search_term})"
                    }
                ]
            }
        },
        "PluginStates": {
            "jira": {
                "Enable": true
            },
            "mattermost-autolink": {
                "Enable": true
            }

        }
    }
```

**Comment by towolf** [comment]
@lieut-data see above ...

**Comment by lieut-data** [comment]
Thanks @towolf, that should definitely help us pin it down. @amyblais, can we put this into triage as a ticket, likely for ABC to investigate?

**Comment by amyblais** [comment]
Opened a ticket here: https://mattermost.atlassian.net/browse/MM-13678.

**Comment by towolf** [comment]
BTW, we use Ubuntu 16.04...

**Comment by towolf** [comment]
systemd service file:
```ini
[Unit]
Description=Mattermost
After=syslog.target network.target mysql.service
Requires=mysql.service
Wants=nginx.service

[Service]
Type=notify
User=mattermost
Group=mattermost
WorkingDirectory=/opt/mattermost
ExecStart=/opt/mattermost/bin/mattermost --disableconfigwatch --config gameforge-config.json
LimitNOFILE=49152
PrivateTmp=true
PrivateDevices=true
Restart=always
RestartSec=10
### new setting: KillMode=process

[Install]
WantedBy=multi-user.target
```

**Comment by lieut-data** [comment]
@towolf, we've found and fixed an issue that I suspect is behind your report: https://github.com/mattermost/mattermost-server/pull/10081.

I'm going to discuss getting this into v5.7 at a minimum, and perhaps you can help us verify in your environment?

**Comment by lieut-data** [comment]
Heads up that we're going to keep this in `master` only for now for release in v5.8 (February 2018).

**Comment by towolf** [comment]
@lieut-data Sure, I'll follow the release notes and report back. In February then, yes?

**Comment by amyblais** [comment]
Closing this as this has been fixed for 5.8 which will be available by tomorrow,
